src/Common/Services/Mail/MailProviderFactory.cs

```diff
@@ -0,0 +1,42 @@
+using Passwordless.Common.Services.Mail.Aws;
+using Passwordless.Common.Services.Mail.File;
+using Passwordless.Common.Services.Mail.SendGrid;
+using Passwordless.Common.Services.Mail.Smtp;
+
+namespace Passwordless.Common.Services.Mail;
+
+public class MailProviderFactory : IMailProviderFactory
+{
+    private readonly IServiceProvider _serviceProvider;
+
+    public MailProviderFactory(IServiceProvider serviceProvider)
+    {
+        _serviceProvider = serviceProvider;
+    }
+
+    public IMailProvider Create(string name, BaseMailProviderOptions options)
+    {
+        switch (name)
+        {
+            case AwsMailProviderOptions.Provider:
+                var awsOptions = (AwsMailProviderOptions)options;
+                var awsMailProviderLogger = _serviceProvider.GetRequiredService<ILogger<AwsMailProvider>>();
+                return new AwsMailProvider(awsOptions, awsMailProviderLogger);
```

Comment:
Why not register the providers as services and let `IServiceProvider` handle dependencies?